core/state/state_witness.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	for codeHash, code := range w.codes {
+		codeHashes = append(codeHashes, codeHash)
+		codes = append(codes, code)
+	}
+
+	for blockNum, blockHash := range w.blockHashes {
+		blockNums = append(blockNums, blockNum)
+		blockHashes = append(blockHashes, blockHash)
+	}
+	l := eb.List()
+	w.block.EncodeRLP(eb)
+	eb.WriteBytes(root[:])
+	if err := rlp.Encode(eb, owners); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, allPaths); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, allNodes); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, blockNums); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, blockHashes); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, codeHashes); err != nil {
+		panic(err)
+	}
+	if err := rlp.Encode(eb, codes); err != nil {
+		panic(err)
+	}
+	eb.ListEnd(l)
+	eb.Flush()
+	return buf.Bytes()
+}
+
+func (w *Witness) addAccessList(owner common.Hash, list map[string][]byte) {
+	if len(list) > 0 {
+		w.lists[owner] = list
+	}
+}
+
+func (w *Witness) AddBlockHash(hash common.Hash, num uint64) {
+	w.blockHashes[num] = hash
+}
+
+// TODO: don't include the code hash in the witness if not necessary
+func (w *Witness) AddCode(hash common.Hash, code Code) {
+	if code, ok := w.codes[hash]; ok && len(code) > 0 {
+		return
+	}
+	w.codes[hash] = code
+}
+
+func (w *Witness) AddCodeHash(hash common.Hash) {
+	if _, ok := w.codes[hash]; ok {
+		return
+	}
+	w.codes[hash] = []byte{}
+}
+
+func (w Witness) Copy() Witness {
+	panic("not implemented")
+}
+
+func (w *Witness) LogSizeWithBlock(b *types.Block) {
+	enc := w.EncodeRLP()
+	fmt.Printf("block %d witness+block size: %d\n", b.Number(), len(enc))
+}
+
+func (w *Witness) Summary() string {
+	b := new(bytes.Buffer)
+	xx, err := rlp.EncodeToBytes(w.block)
+	if err != nil {
+		panic(err)
+	}
+	totBlock := len(xx)
+
+	yy := w.EncodeRLP()
+
+	totWit := len(yy)
+	totCode := 0
+	for _, c := range w.codes {
+		totCode += len(c)
+	}
+	totNodes := 0
+	totPaths := 0
+	nodePathCount := 0
+	for _, ownerPaths := range w.lists {
+		for path, node := range ownerPaths {
+			nodePathCount++
+			totNodes += len(node)
+			totPaths += len(path)
+		}
+	}
+
+	fmt.Fprintf(b, "%4d hashes: %v\n", len(w.blockHashes), common.StorageSize(len(w.blockHashes)*32))
+	fmt.Fprintf(b, "%4d owners: %v\n", len(w.lists), common.StorageSize(len(w.lists)*32))
+	fmt.Fprintf(b, "%4d nodes:  %v\n", nodePathCount, common.StorageSize(totNodes))
+	fmt.Fprintf(b, "%4d paths:  %v\n", nodePathCount, common.StorageSize(totPaths))
+	fmt.Fprintf(b, "%4d codes:  %v\n", len(w.codes), common.StorageSize(totCode))
+	fmt.Fprintf(b, "%4d codeHashes: %v\n", len(w.codes), common.StorageSize(len(w.codes)*32))
+	fmt.Fprintf(b, "block (%4d txs): %v\n", len(w.block.Transactions()), common.StorageSize(totBlock))
+	fmt.Fprintf(b, "Total size: %v\n ", common.StorageSize(totWit))
+	return b.String()
+}
+
+func (w *Witness) PopulateMemoryDB() ethdb.Database {
+	db := rawdb.NewMemoryDatabase()
```

Comment:
Maybe we should  use a special memory db here, one which screams loudly whenever a `get`-operation fails. Because that indicates that the witness is lacking, no? 

Could just do a `ethdb.Database` middle layer which logs misses.